### PR TITLE
Poll module

### LINKED ISSUE #44: Voting Module
Module that will allow campaigns to be created and users can vote on the campaign. On starting a campaign the bot will post a message to the specified channel with reactions as the different options.

Campaigns will be added to the database and tracked from the website. Votes will be tallied and tracked in the database based on reactions to the original message.

### PR DESCRIPTION
**Pull request by consolelogreece**
Note - Currently I manually validate that a user is posting to a channel that belongs to their guild. This should probably be extracted into a web filter.

Closes #44

### CODE REVIEW

File: src/Volvox.Helios.Web/Views/Poll/NewPoll.cshtml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+            <div class="form-group">
+                <label asp-for="Title"></label> <br />
+                <input asp-for="Title" type="text" class="form-control" />
+                <span asp-validation-for="Title" style="color:red"></span>
+            </div>
+
+            <input type="hidden" asp-for="TotOptions" value="@Model.TotOptions" />
+
+            @for (int i = 0; i < Model.TotOptions; i++)
```
Comment (BillChirico) [inline src/Volvox.Helios.Web/Views/Poll/NewPoll.cshtml]: `TotalOptions`

File: src/Volvox.Helios.Web/Controllers/PollController.cs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+            if(!ModelState.IsValid) return View(viewModel);
+
+            var poll = new StringBuilder();
+
+            // Add poll title.
+            poll.Append(":small_blue_diamond: " + viewModel.Title + Environment.NewLine);
+
+            // Unicode representations of discord's 0-10 emoticons.
+            var discordNumbers = new string[11] { "0\u20e3", "1\u20e3", "2\u20e3", "3\u20e3", "4\u20e3", "5\u20e3", "6\u20e3", "7\u20e3", "8\u20e3", "9\u20e3", "🔟" };
```
Comment (BillChirico) [inline src/Volvox.Helios.Web/Controllers/PollController.cs]: Move these to constants

File: src/Volvox.Helios.Web/Controllers/ApiController.cs [hunk 3]
```diff
@@ -72,5 +79,47 @@ public bool IsBotInGuild(ulong guildId, [FromServices] IBot bot)
...
+
+            // Splitting by newline will provide array with title at index 0, and the rest will be the options
+            var pollDetails = poll.Content.Split(Environment.NewLine);
+
+            formattedPoll.Title = pollDetails[0];
+
+            var pollOptions = new List<OptionModel>();
+
+            // Unicode representations of discord 0-10 emoticons.
+            var discordNumbers = new string[11] { "0\u20e3", "1\u20e3", "2\u20e3", "3\u20e3", "4\u20e3", "5\u20e3", "6\u20e3", "7\u20e3", "8\u20e3", "9\u20e3", "🔟" };
```
Comment (BillChirico) [inline src/Volvox.Helios.Web/Controllers/ApiController.cs]: Move to constant

File: src/Volvox.Helios.Domain/Module/Poll.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+using System.ComponentModel.DataAnnotations;
+using System.Text;
+using Volvox.Helios.Domain.ModuleSettings;
+
+namespace Volvox.Helios.Domain.Module
+{
+    public class Poll
+    {
+        [Key]
+        public ulong MessageId { get; set; }
```
Comment (mhallmark) [inline src/Volvox.Helios.Domain/Module/Poll.cs]: The class name here is "Poll" but the primary key is "MessageId". Typical convention is to just use "Id" as the main identifier for a class/table, or an alternative "PollId" would suffice here also

File: src/Volvox.Helios.Web/Controllers/PollController.cs [hunk 5]
```diff
@@ -37,19 +41,58 @@ public async Task<IActionResult> NewPoll(ulong guildId, [FromServices] IDiscordG
...
+
+            // Add options to poll.
+            for (var i = 0; i < validOptions.Count; i++)
+                poll.Append($"{discordNumbers[i + 1]} {validOptions[i]}{Environment.NewLine}");
+
+            // Post message and get message details.
+            var message = await messageService.Post(viewModel.ChannelId, poll.ToString());
```
Comment (mhallmark) [inline src/Volvox.Helios.Web/Controllers/PollController.cs]: Leaving for future note - EntityServices and ModuleSettingsServices will be exposing changed events. Modules should subscribe to those data changed events to update or post messages and leave the controller to only have to worry about changing the underlying data.

File: src/Volvox.Helios.Web/Controllers/PollController.cs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+            return RedirectToAction("Index");
+        }
+        #endregion
+
+        #region ViewPoll
+        [HttpGet]
+        public async Task<IActionResult> Index(ulong guildId, [FromServices] IEntityService<Poll> entityServicePolls)
+        {
+            var polls = await entityServicePolls.Get(p => p.GuildId == guildId);
```
Comment (GusTheGray) [inline src/Volvox.Helios.Web/Controllers/PollController.cs]: Who should be able to view this data? Just the person that setup the poll? Administrators for the guild? Or anyone?

**Pull request merged**